pkg/upgrade/config/v1alpha1/config.go

```diff
@@ -0,0 +1,199 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+)
+
+// UpgradeConfig is wrapper over apis.UpgradeConfig which is
+// upgrade config for a particular job.
+type UpgradeConfig struct {
+	object *apis.UpgradeConfig
+}
+
+// Builder helps to build UpgradeConfig instance
+type Builder struct {
+	*UpgradeConfig
+	checks map[*Predicate]string
+	errors []error
+}
+
+// Predicate abstracts conditional logic w.r.t the upgradeConfig instance
+//
+// NOTE:
+// Predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// Predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type Predicate func(*UpgradeConfig) bool
+
+// AddCheckf adds the predicate as a condition to be validated against the
+// upgrade config instance and format the message string according to format specifier.
+// If only predicate and message string is provided, it will treat it as the
+// value for the corresponding predicate.
+func (b *Builder) AddCheckf(p Predicate, predicateMsg string, args ...interface{}) *Builder {
+	b.checks[&p] = fmt.Sprintf(predicateMsg, args...)
+	return b
+}
+
+// AddCheck adds the predicate as a condition to be validated against the
+// upgrade config instance
+func (b *Builder) AddCheck(p Predicate) *Builder {
+	return b.AddCheckf(p, "")
+}
+
+// AddChecks adds the provided predicates as conditions to be validated against
+// the upgrade config instance
+func (b *Builder) AddChecks(predicates ...Predicate) *Builder {
+	for _, check := range predicates {
+		b.AddCheck(check)
+	}
+	return b
+}
+
+// Build returns the final instance of UpgradeConfig
+func (b *Builder) Build() (*apis.UpgradeConfig, error) {
+	if len(b.errors) != 0 {
+		return nil, errors.Errorf("%v", b.errors)
+	}
+	err := b.validate()
+	if err != nil {
+		return nil, err
+	}
+	return b.UpgradeConfig.object, nil
+}
+
+// validate will run checks against UpgradeConfig instance
+func (b *Builder) validate() error {
+	for cond := range b.checks {
+		pass := (*cond)(b.UpgradeConfig)
+		if !pass {
+			b.errors = append(b.errors,
+				errors.Errorf("%v", b.checks[cond]))
+		}
+	}
+	if len(b.errors) == 0 {
+		return nil
+	}
+	return errors.Errorf("%v", b.errors)
+}
+
+// NewBuilder returns a new instance of Builder
+func NewBuilder() *Builder {
+	return &Builder{
+		UpgradeConfig: &UpgradeConfig{
+			object: &apis.UpgradeConfig{},
+		},
+		checks: make(map[*Predicate]string),
+	}
+}
+
+// WithYamlString add object in builder struct.
+func (b *Builder) WithYamlString(yamlString string) *Builder {
```

Comment:
This should be a variant of Builder instantiation

Reply:
done